### ISSUE TITLE
receive 401 error for backing up couchDB Operator deployment on OpenShift 3.11

### ISSUE THREAD

**Issue by byroncollins** [cloudant/couchbackup#334]
## Bug Description

We used the IBM Kubernetes couchdb operator to deploy a couchdb cluster onto our OpenShift 3.11 cluster.

Operator: https://operatorhub.io/operator/couchdb-operator
Documentation: https://cloud.ibm.com/docs/Cloudant?topic=Cloudant-installing-the-operator#installing-operator-openshift

The couchdb deploys successfully and we can interact as expected using curl.

```bash
[~] curl https://*****:*****@my-couchdb-byronc-couchdb.apps.dev.containers.datacom.cloud
{"couchdb":"Welcome","version":"2.3.1","git_sha":"c298091a4","uuid":"bb26c57d-a6d3-11ea-8909-0a580a81001c","features":["pluggable-storage-engines","scheduler"],"vendor":{"name":"IBM"}}
```
We explored using couchbackup to backup databases running on the cluster.

```bash
curl https://***:***@my-couchdb-byronc-couchdb.apps.dev.containers.datacom.cloud/couchbackup-demo
{"db_name":"couchbackup-demo","purge_seq":"0-g1AAAACDeJzLYWBgYMpgTmFQT84vTc5ISXJI1s2t1IVydHN1DfVQBXJAGvJYgCRDA5D6DwRZiQwkm5DIkFQP0ZoFAEq9KnI","update_seq":"11-g1AAAACDeJzLYWBgYMpgTmFQT84vTc5ISXJI1s2t1IVydHN1DfVQBXJAGhIZkur___-flchKstY8FiDJ0ACk_oNNYMsCAJcmKn0","sizes":{"file":74174,"external":707,"active":3736},"other":{"data_size":707},"doc_del_count":0,"doc_count":5,"disk_size":74174,"disk_format_version":7,"data_size":3736,"compact_running":false,"cluster":{"q":2,"n":3,"w":2,"r":2},"instance_start_time":"0"}
```
but when we execute the couchbackup command we get back the Error: Failed to get cookie. Status code: 401

```bash
[~]  couchbackup --url https://****:****@my-couchdb-byronc-couchdb.apps.dev.containers.datacom.cloud --db couchbackup-demo
================================================================================
Performing backup on https://****:****@my-couchdb-byronc-couchdb.apps.dev.containers.datacom.cloud/couchbackup-demo using configuration:
{
  "bufferSize": 500,
  "log": "/tmp/tmp-82060PV8HSHSMcUd",
  "mode": "full",
  "parallelism": 5,
  "requestTimeout": 120000,
  "resume": false
}
================================================================================
  couchbackup:backup Fetching all database changes... +0ms
(node:8206) UnhandledPromiseRejectionWarning: Error: Failed to get cookie. Status code: 401
    at Request._callback (/usr/lib/node_modules/@cloudant/couchbackup/node_modules/@cloudant/cloudant/lib/tokens/CookieTokenManager.js:48:18)
    at Request.self.callback (/usr/lib/node_modules/@cloudant/couchbackup/node_modules/request/request.js:185:22)
    at Request.emit (events.js:315:20)
    at Request.<anonymous> (/usr/lib/node_modules/@cloudant/couchbackup/node_modules/request/request.js:1154:10)
    at Request.emit (events.js:315:20)
    at IncomingMessage.<anonymous> (/usr/lib/node_modules/@cloudant/couchbackup/node_modules/request/request.js:1076:12)
    at Object.onceWrapper (events.js:421:28)
    at IncomingMessage.emit (events.js:327:22)
    at endReadableNT (_stream_readable.js:1224:12)
    at processTicksAndRejections (internal/process/task_queues.js:84:21)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:8206) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:8206) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```
### 2. What you expected to happen

That the database is backed up as expected by couchbackup

### 3. What actually happened
`
Error: Failed to get cookie. Status code: 401

Not sure if it's an issue with the operator deployment of couchdb or an issue with the couchbackup tool.
`
## Environment details

I've tested with the same results on on both ubuntu 18.04 (Node v14.4.0) and Centos 8 (Node v10.19.0) against the same couchdb cluster.

- Version(s) that are affected by this issue.
    > 2.3.1
- Node.js version
    > v14.4.0
   > v10.19.0

**Comment by ricellis** [comment]
This is caused by the CouchDB setting `require_valid_user`. The operator docs state:

>Authentication - The parameter require_valid_user defaults to true, which means that no requests are allowed from anonymous users. Every request must be authenticated.

I see that you are using Couch 2.3.1 - in Couch 2.x if this option is enabled the credentials passed in the `_session` request body are not used to authorize the user so it is not possible to obtain the cookie without additionally providing the credentials in a Basic auth header (the underlying libraries used here don't do that). I'd recommend updating to Couch 3.x where [this problem is resolved](https://github.com/apache/couchdb/issues/2321).

**Comment by welshDoug** [comment]
Support for version 2.x of CouchDB in the Operator is being deprecated in a forthcoming release (no specific date yet) so we recommend that customers upgrade to version 3.x

The operator supports the latest version of CouchDB 3.x currently 3.1.0. If you update your version property in the CouchDBCluster resource to either 3.0.0 or 3.1.0 the operator will deploy CouchDB version 3.1.0

In order to update the pods they will need to be reloaded, you can do this by manually deleting a pod. It is advisable to wait for the new pod to come up before moving to the next one.

**Comment by byroncollins** [comment]
This is resolved when running against a CouchDB version 3.1.0 deployed by the Operator
